piptools/scripts/compile.py

```diff
@@ -421,7 +427,13 @@ def cli(
 
     # Filter out pip environment markers which do not match (PEP496)
     constraints = [
-        req for req in constraints if req.markers is None or req.markers.evaluate()
+        req
+        for req in constraints
+        if req.markers is None
+        # We explicitly set extra=None to filter out optional requirements
+        # since evaluating an extra marker with no environment raises UndefinedEnvironmentName
+        # (see https://packaging.pypa.io/en/latest/markers.html#usage)
+        or req.markers.evaluate(dict(extra=None))
```

Comment:
```suggestion
        or req.markers.evaluate({"extra": None})
```

We could use literals - they are faster.